src/frontend/src/optimizer/plan_visitor/cardinality_visitor.rs

```diff
@@ -0,0 +1,402 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+use std::ops::{Mul, Sub};
+
+use risingwave_pb::plan_common::JoinType;
+
+use self::card::Cardinality;
+use super::{DefaultBehavior, DefaultValue, PlanVisitor};
+use crate::catalog::system_catalog::pg_catalog::PG_NAMESPACE_TABLE_NAME;
+use crate::optimizer::plan_node::generic::Limit;
+use crate::optimizer::plan_node::{self, PlanTreeNode, PlanTreeNodeBinary, PlanTreeNodeUnary};
+use crate::optimizer::plan_visitor::PlanRef;
+
+pub mod card {
+    use std::cmp::{max, min, Ordering};
+    use std::ops::{Add, Mul, Sub};
+
+    /// The upper bound of the [`Cardinality`].
+    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
+    pub enum Hi {
+        Limited(usize),
+        Unlimited,
+    }
+
+    impl PartialOrd for Hi {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            Some(self.cmp(other))
+        }
+    }
+
+    impl Ord for Hi {
+        fn cmp(&self, other: &Self) -> Ordering {
+            match (self, other) {
+                (Self::Unlimited, Self::Unlimited) => Ordering::Equal,
+                (Self::Unlimited, Self::Limited(_)) => Ordering::Greater,
+                (Self::Limited(_), Self::Unlimited) => Ordering::Less,
+                (Self::Limited(lhs), Self::Limited(rhs)) => lhs.cmp(rhs),
+            }
+        }
+    }
+
+    impl From<Option<usize>> for Hi {
+        fn from(value: Option<usize>) -> Self {
+            value.map_or(Self::Unlimited, Self::Limited)
+        }
+    }
+
+    impl From<usize> for Hi {
+        fn from(value: usize) -> Self {
+            Self::Limited(value)
+        }
+    }
+
+    /// The cardinality of the output rows of a plan node. Bounds are inclusive.
+    ///
+    /// The default value is `0..`, i.e. the number of rows is unknown.
+    // TODO: Make this the property of each plan node.
+    #[derive(Clone, Copy, Debug)]
+    pub struct Cardinality {
```

Comment:
Why not simply use `std::ops::RangeBounds` (i.e., `lo..hi`s) 😄

Reply:
It's either `RangeInclusive` and `RangeFrom`. By using an enum, I guess the code complexity can be similar. 🥵